File: apps/api/src/rag/retriever.ts
import type { RagTopic, VectorHit } from './vectorStore';
import { RAG_TOPICS, vectorStore } from './vectorStore';

/** Dokumen yang diambil sebagai konteks jawaban. */
export interface RetrievedDoc {
  file: string;
  section?: string;
  snippet?: string;
  distance?: number;
}

export interface StoreLike {
  query(topic: RagTopic, embedding: number[], topK: number): Promise<VectorHit[]>;
}

export interface RetrieverOptions {
  store?: StoreLike;
  topics?: readonly RagTopic[];
  topK?: number;
}

/**
 * Retrieval cosine-similarity dari seluruh collection topic,
 * digabung lalu diambil top_k teratas berdasarkan distance (kecil = mirip).
 */
export async function retrieve(
  embedding: number[],
  options: RetrieverOptions = {},
): Promise<RetrievedDoc[]> {
  const store = options.store ?? vectorStore;
  const topics = options.topics ?? RAG_TOPICS;
  const topK = options.topK ?? 0; // caller pas; default kosong bila 0

  const perTopic = Math.max(1, Math.ceil(topK / topics.length));

  const results = await Promise.all(
    topics.map(async (topic) => {
      try {
        return await store.query(topic, embedding, perTopic);
      } catch {
        return [] as VectorHit[];
      }
    }),
  );

  const merged = flatMap(results)
    .filter((h) => h.text.length > 0)
    .sort((a, b) => a.distance - b.distance);

  const taken = topK > 0 ? merged.slice(0, topK) : merged;

  return taken.map((h) => ({
    file:
      (typeof h.metadata.file === 'string' && h.metadata.file) ||
      String(h.id.split('#')[0] ?? 'rag'),
    section: typeof h.metadata.section === 'string' ? h.metadata.section : undefined,
    snippet: h.text,
    distance: h.distance,
  }));
}

function flatMap(hits: VectorHit[][]): VectorHit[] {
  return hits.reduce((acc, h) => acc.concat(h), [] as VectorHit[]);
}

export { RAG_TOPICS };
export type { RagTopic, VectorHit };

File: apps/ai/app/rag/retriever.py
"""Retriever: cosine-similarity across all topic collections, merged, top_k.
Port of `apps/api/src/rag/retriever.ts`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from app.rag.vector_store import RAG_TOPICS, VectorStore, vector_store


@dataclass
class RetrievedDoc:
    file: str
    section: Optional[str] = None
    snippet: Optional[str] = None
    distance: Optional[float] = None


class StoreLike:
    """Protocol for a fake store in tests."""

    async def query(self, topic: str, embedding: list[float], top_k: int) -> list[dict]:
        raise NotImplementedError


class Retriever:
    def __init__(self, store: VectorStore | StoreLike = vector_store, topics: Optional[list[str]] = None) -> None:
        self.store = store
        self.topics = topics if topics is not None else list(RAG_TOPICS)

    async def retrieve(
        self,
        embedding: list[float],
        top_k: int = 5,
    ) -> list[RetrievedDoc]:
        per_topic = max(1, (top_k + len(self.topics) - 1) // len(self.topics)) if top_k else 1

        results: list[list[dict]] = []
        for topic in self.topics:
            try:
                hits = await self.store.query(topic, embedding, per_topic)
                results.append(hits)
            except NotImplementedError:
                raise
            except Exception:
                results.append([])

        merged: list[dict] = []
        for hits in results:
            for h in hits:
                if h.get("text"):
                    merged.append(h)
        merged.sort(key=lambda h: h.get("distance", 0.0))

        taken = merged[:top_k] if top_k > 0 else merged

        docs: list[RetrievedDoc] = []
        for h in taken:
            meta = h.get("metadata") or {}
            file = meta.get("file") or str(h.get("id", "")).split("#")[0] or "rag"
            docs.append(
                RetrievedDoc(
                    file=str(file),
                    section=meta.get("section") if isinstance(meta.get("section"), str) else None,
                    snippet=h.get("text"),
                    distance=h.get("distance"),
                )
            )
        return docs


retriever = Retriever()